File: frontend/src/lib/feeUtils.js
export const calculateExpectedFee = (contract, totalAssets) => {
    if (!contract) {
      return {
        fee: null,
        method: 'No contract data',
      };
    }
    
    if (contract.fee_type === 'flat') {
      return {
        fee: contract.flat_rate,
        method: 'Flat fee',
      };
    }
    
    if (contract.fee_type === 'percentage' || contract.fee_type === 'percent') {
      if (!totalAssets || !contract.percent_rate) {
        return {
          fee: null,
          method: 'Percentage fee (missing data)',
        };
      }
      
      const fee = totalAssets * contract.percent_rate;
      const ratePercentage = (contract.percent_rate * 100).toFixed(4);
      
      return {
        fee,
        method: `${ratePercentage}% of $${formatCurrency(totalAssets)}`,
      };
    }
    
    return {
      fee: null,
      method: 'Unknown fee type',
    };
  };
  
  export const formatCurrency = (value) => {
    if (value === null || value === undefined) return 'N/A';
    
    return new Intl.NumberFormat('en-US', {
      style: 'currency',
      currency: 'USD',
      minimumFractionDigits: 2,
      maximumFractionDigits: 2,
    }).format(value);
  };
  
  export const formatPercentage = (value) => {
    if (value === null || value === undefined) return 'N/A';
    
    return new Intl.NumberFormat('en-US', {
      style: 'percent',
      minimumFractionDigits: 2,
      maximumFractionDigits: 2,
    }).format(value / 100);
  };
  
  export const generateFeeReferences = (contract) => {
    if (!contract) return null;
    
    let monthlyRate, quarterlyRate, annualRate;
    
    if (contract.fee_type === 'flat') {
      if (contract.payment_schedule === 'monthly') {
        monthlyRate = contract.flat_rate;
        quarterlyRate = monthlyRate * 3;
        annualRate = monthlyRate * 12;
      } else {
        quarterlyRate = contract.flat_rate;
        monthlyRate = quarterlyRate / 3;
        annualRate = quarterlyRate * 4;
      }
    } else if (contract.percent_rate) {
      monthlyRate = contract.payment_schedule === 'monthly' 
        ? contract.percent_rate 
        : contract.percent_rate / 3;
      
      quarterlyRate = contract.payment_schedule === 'monthly'
        ? contract.percent_rate * 3
        : contract.percent_rate;
        
      annualRate = contract.payment_schedule === 'monthly'
        ? contract.percent_rate * 12
        : contract.percent_rate * 4;
    } else {
      return null;
    }
    
    return {
      monthly: contract.fee_type === 'flat' ? formatCurrency(monthlyRate) : (monthlyRate * 100).toFixed(4) + '%',
      quarterly: contract.fee_type === 'flat' ? formatCurrency(quarterlyRate) : (quarterlyRate * 100).toFixed(4) + '%',
      annual: contract.fee_type === 'flat' ? formatCurrency(annualRate) : (annualRate * 100).toFixed(4) + '%',
    };
  };

File: tests/utils/fee_utils.py
# tests/utils/fee_utils.py
def format_currency(value):
    """Format a numerical value as currency."""
    if value is None:
        return "N/A"
    return f"${value:,.2f}"

def generate_fee_references(contract):
    """
    Generate fee references for different time periods based on the contract.
    This is a Python implementation of the frontend's generateFeeReferences.
    
    Parameters:
    contract (dict): Contract details including fee_type, flat_rate, percent_rate, and payment_schedule
    
    Returns:
    dict: Monthly, quarterly, and annual fee references
    """
    if not contract:
        return None
    
    monthly_rate, quarterly_rate, annual_rate = None, None, None
    
    if contract.get("fee_type") == "flat":
        if contract.get("payment_schedule") == "monthly":
            monthly_rate = contract.get("flat_rate")
            quarterly_rate = monthly_rate * 3 if monthly_rate else None
            annual_rate = monthly_rate * 12 if monthly_rate else None
        else:  # quarterly
            quarterly_rate = contract.get("flat_rate")
            monthly_rate = quarterly_rate / 3 if quarterly_rate else None
            annual_rate = quarterly_rate * 4 if quarterly_rate else None
    elif contract.get("percent_rate"):
        if contract.get("payment_schedule") == "monthly":
            monthly_rate = contract.get("percent_rate")
            quarterly_rate = monthly_rate * 3 if monthly_rate else None
            annual_rate = monthly_rate * 12 if monthly_rate else None
        else:  # quarterly
            quarterly_rate = contract.get("percent_rate")
            monthly_rate = quarterly_rate / 3 if quarterly_rate else None
            annual_rate = quarterly_rate * 4 if quarterly_rate else None
    else:
        return None
    
    return {
        "monthly": format_currency(monthly_rate) if contract.get("fee_type") == "flat" else f"{monthly_rate * 100:.4f}%" if monthly_rate else None,
        "quarterly": format_currency(quarterly_rate) if contract.get("fee_type") == "flat" else f"{quarterly_rate * 100:.4f}%" if quarterly_rate else None,
        "annual": format_currency(annual_rate) if contract.get("fee_type") == "flat" else f"{annual_rate * 100:.4f}%" if annual_rate else None,
    }